crypto/statetrie/nibbles.go

```diff
@@ -0,0 +1,162 @@
+// Copyright (C) 2019-2023 Algorand, Inc.
+// This file is part of go-algorand
+//
+// go-algorand is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as
+// published by the Free Software Foundation, either version 3 of the
+// License, or (at your option) any later version.
+//
+// go-algorand is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with go-algorand.  If not, see <https://www.gnu.org/licenses/>.
+
+package statetrie
+
+import (
+	"bytes"
+	"errors"
+)
+
+// Nibbles are 4-bit values stored in an 8-bit byte arrays
+type Nibbles []byte
+
+// MakeNibbles returns a nibble array from the byte array.  If half is true, the
+// last 4 bits of the last byte of the array are ignored.
+func MakeNibbles(data []byte, half bool) Nibbles {
+	return unpack(data, half)
+}
+
+// Unpack the byte array into a nibble array.  If half is true, the last 4
+// bits of the last byte of the array are ignored.  Allocates a new byte
+// slice.
+//
+// [0x12, 0x30], true -> [0x1, 0x2, 0x3]
+// [0x12, 0x34], false -> [0x1, 0x2, 0x3, 0x4]
+// [0x12, 0x34], true -> [0x1, 0x2, 0x3]  <-- last byte last 4 bits ignored
+// [], false -> []
+// never to be called with [], true
+func unpack(data []byte, half bool) Nibbles {
```

Comment:
nit: also a bit opinionated on `half` naming tho, for returning nibbles length would be always odd if `half` being true, maybe `dropLast` or `oddLength` would be easier to understand? just saying.

Reply:
[#93d95](https://github.com/algorand/go-algorand/pull/5719/commits/93d95b502408fe37f9435714efeaffe67cc0ffd9)